proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,260 @@
+While this example shows only one version where of forward compatibility (when Y and Y+1 are supported, but Y is still written), in practice there could be more "forward compatible releases" within this strategy.
+
+Technically we have to keep old format read (and write) support until the next major release.
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise.
+  * **Mitigation**: We also plan guide and documentation.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner.
+
+## Alternatives
+
+### Require LTS support
+
+We could add a variation to the [Two-Fold Migration Strategy](#two-fold-migration-strategy) (let's call it a "LTS migration strategy") where both X+1 and the last LTS (long time support version) before X+1, is able to read Y+1 version. Only then we are allowed to release X+2 that switches the default.
+
+For example:
+
+1. LTS 3.1 only supports WAL v1.
+2. 3.3 adds WAL v2.
+3. we wait unit next LTS so e.g. 3.24.
+4. 3.25 can now switch to WAL v2.
+5. 4.0 can remove WAL v1 support.
+
+Pros:
+* Gives a bit more stability to users and less surprises.
+
+Cons:
+* Extremely heavy process that will make us afraid/refuse to make improvements to WAL, because it's too much work. It might fails our goal of `Balancing development velocity with user data stability risks`
+  * One mitigation would be an LTS retroactive strategy e.g. LTS 3.1 only supports WALv1, 3.3 adds WALv2, we do 3.1.1 with WALv2 too, 3.4 can now switch to WALv2, 4.0 can remove WALv1 support. It gives us more flexibility, but it's not very realistic to do patch release of LTS with risky feature like a new WAL.
+* We literally have no formal process for LTS versions and we don't do them regularly.
+
+### Add a CLI tool that rewrites WAL to a specific version
+
+We could add a `promtool` command that rewrites WAL segments to a given version.
+
+Users can then migrate their WAL with a single command either as an init process before reverting or manually on a disk.
+
+This should take ~minutes even for larger WAL files.
+
+Pros:
+* Less need for a two-fold strategy?
+
+Cons:
+* We need to write migration code and ensure it's efficient enough for bigger data
+* A bit painful to use on scale and remotely (e.g. on Kubernetes)
+
+### Rewrite WAL before/during replay
+
+Instead of supporting multiple directories for WAL for various versions, we could rewrite WAL on the start.
+
+Pros:
+* Simpler implementation
+
+Cons:
+* We already suffer from replay problems, so I propose an eventual rewrite.
+* Rewrite is more risking than read-only WAL (of previous version)
+
+### Don't version WAL, don't introduce a flag
```

Comment:
```suggestion
### Alternative: Don't version WAL, don't introduce a flag
```
I suggest to put "Alternative" in each of these sub-headings, because it isn't clear once you've scrolled down a bit.

Reply:
E_TOO_MANY_ALTERNATIVES